File: saucelabs-selenium-project/utils/driver_factory.py
"""
driver_factory.py - Creates and manages WebDriver instances.
Supports local and Sauce Labs remote execution.
Uses Selenium's built-in driver manager (no SSL issues) with
webdriver-manager as fallback.
"""
import os
import ssl
from selenium import webdriver
from selenium.webdriver.chrome.service import Service as ChromeService
from selenium.webdriver.firefox.service import Service as FirefoxService
from selenium.webdriver.edge.service import Service as EdgeService
from utils.config_reader import ConfigReader

# Disable SSL verification globally for environments with proxy/firewall issues
os.environ["WDM_SSL_VERIFY"] = "0"


class DriverFactory:
    """Factory class to create WebDriver instances for local or Sauce Labs."""

    @staticmethod
    def create_driver(test_name="Unnamed Test"):
        """Create and return a WebDriver instance based on configuration."""
        if ConfigReader.is_sauce_enabled():
            return DriverFactory._create_sauce_driver(test_name)
        return DriverFactory._create_local_driver()

    @staticmethod
    def _create_local_driver():
        """Create a local WebDriver instance using Selenium's built-in driver manager."""
        browser = ConfigReader.get_browser().lower()
        headless = ConfigReader.is_headless()

        if browser == "firefox":
            options = webdriver.FirefoxOptions()
            if headless:
                options.add_argument("--headless")
            driver = webdriver.Firefox(options=options)

        elif browser == "edge":
            options = webdriver.EdgeOptions()
            if headless:
                options.add_argument("--headless")
            driver = webdriver.Edge(options=options)

        else:  # default: chrome
            options = webdriver.ChromeOptions()
            if headless:
                options.add_argument("--headless=new")
            options.add_argument("--disable-notifications")
            options.add_argument("--start-maximized")
            driver = webdriver.Chrome(options=options)

        driver.implicitly_wait(ConfigReader.get_implicit_wait())
        driver.set_page_load_timeout(ConfigReader.get_page_load_timeout())
        driver.maximize_window()
        return driver

    @staticmethod
    def _create_sauce_driver(test_name):
        """Create a Sauce Labs RemoteWebDriver instance."""
        username = ConfigReader.get_sauce_username()
        access_key = ConfigReader.get_sauce_access_key()

        sauce_options = {
            "username": username,
            "accessKey": access_key,
            "name": test_name,
            "build": ConfigReader.get("saucelabs", "build_name", fallback="Local Build"),
            "screenResolution": ConfigReader.get("saucelabs", "screen_resolution", fallback="1920x1080"),
            "maxDuration": 3600,
            "idleTimeout": 300,
        }

        browser_name = ConfigReader.get("saucelabs", "browser_name", fallback="chrome")

        if browser_name == "firefox":
            options = webdriver.FirefoxOptions()
        elif browser_name == "edge":
            options = webdriver.EdgeOptions()
        else:
            options = webdriver.ChromeOptions()

        options.browser_version = ConfigReader.get("saucelabs", "browser_version", fallback="latest")
        options.platform_name = ConfigReader.get("saucelabs", "platform_name", fallback="Windows 11")
        options.set_capability("sauce:options", sauce_options)

        sauce_url = f"https://{username}:{access_key}@ondemand.us-west-1.saucelabs.com:443/wd/hub"
        driver = webdriver.Remote(command_executor=sauce_url, options=options)

        driver.implicitly_wait(ConfigReader.get_implicit_wait())
        driver.maximize_window()
        return driver

    @staticmethod
    def update_sauce_result(driver, passed):
        """Update the Sauce Labs job result."""
        if ConfigReader.is_sauce_enabled():
            script = "sauce:job-result={}".format("passed" if passed else "failed")
            driver.execute_script(script)

File: saucelabs-java-selenium/src/main/java/com/saucelabs/driver/DriverFactory.java
package com.saucelabs.driver;

import com.saucelabs.config.ConfigReader;
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.MutableCapabilities;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.edge.EdgeDriver;
import org.openqa.selenium.edge.EdgeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.remote.RemoteWebDriver;

import java.net.MalformedURLException;
import java.net.URL;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

/**
 * DriverFactory - Creates and manages WebDriver instances.
 * Supports local execution and Sauce Labs remote execution.
 */
public class DriverFactory {

    /**
     * Create and return a WebDriver instance based on configuration.
     */
    public static WebDriver createDriver(String testName) {
        if (ConfigReader.isSauceEnabled()) {
            return createSauceDriver(testName);
        }
        return createLocalDriver();
    }

    /**
     * Create a local WebDriver instance.
     * For Edge/Chrome, uses the locally cached Selenium-manager driver to avoid
     * network downloads in restricted environments.
     */
    private static WebDriver createLocalDriver() {
        String browser = ConfigReader.getBrowser().toLowerCase();
        boolean headless = ConfigReader.isHeadless();
        WebDriver driver;

        switch (browser) {
            case "firefox": {
                setLocalDriverPath("geckodriver", "webdriver.gecko.driver",
                        System.getProperty("user.home") + "/.cache/selenium/geckodriver/win64");
                FirefoxOptions options = new FirefoxOptions();
                if (headless) options.addArguments("--headless");
                driver = new FirefoxDriver(options);
                break;
            }
            case "edge": {
                String edgeDriverPath = findCachedDriver(
                        System.getProperty("user.home") + "\\.cache\\selenium\\msedgedriver\\win64",
                        "msedgedriver.exe");
                if (edgeDriverPath != null) {
                    System.setProperty("webdriver.edge.driver", edgeDriverPath);
                } else {
                    WebDriverManager.edgedriver().setup();
                }
                EdgeOptions options = new EdgeOptions();
                if (headless) options.addArguments("--headless");
                driver = new EdgeDriver(options);
                break;
            }
            default: { // chrome
                String chromeDriverPath = findCachedDriver(
                        System.getProperty("user.home") + "\\.cache\\selenium\\chromedriver\\win64",
                        "chromedriver.exe");
                if (chromeDriverPath != null) {
                    System.setProperty("webdriver.chrome.driver", chromeDriverPath);
                } else {
                    WebDriverManager.chromedriver().setup();
                }
                ChromeOptions options = new ChromeOptions();
                if (headless) options.addArguments("--headless=new");
                options.addArguments("--disable-notifications");
                options.addArguments("--start-maximized");
                driver = new ChromeDriver(options);
                break;
            }
        }

        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(ConfigReader.getImplicitWait()));
        driver.manage().timeouts().pageLoadTimeout(Duration.ofSeconds(ConfigReader.getPageLoadTimeout()));
        driver.manage().window().maximize();
        return driver;
    }

    /**
     * Create a Sauce Labs RemoteWebDriver instance.
     */
    private static WebDriver createSauceDriver(String testName) {
        String username   = ConfigReader.getSauceUsername();
        String accessKey  = ConfigReader.getSauceAccessKey();
        String browserName = ConfigReader.getSauceBrowserName().toLowerCase();

        Map<String, Object> sauceOptions = new HashMap<>();
        sauceOptions.put("username", username);
        sauceOptions.put("accessKey", accessKey);
        sauceOptions.put("name", testName);
        sauceOptions.put("build", ConfigReader.getSauceBuildName());
        sauceOptions.put("screenResolution", ConfigReader.getSauceScreenResolution());
        sauceOptions.put("maxDuration", 3600);
        sauceOptions.put("idleTimeout", 300);

        MutableCapabilities options;
        switch (browserName) {
            case "firefox":
                options = new FirefoxOptions();
                break;
            case "edge":
                options = new EdgeOptions();
                break;
            default:
                options = new ChromeOptions();
                break;
        }

        options.setCapability("browserVersion", ConfigReader.getSauceBrowserVersion());
        options.setCapability("platformName",   ConfigReader.getSaucePlatformName());
        options.setCapability("sauce:options",  sauceOptions);

        String sauceUrl = String.format(
                "https://%s:%s@ondemand.us-west-1.saucelabs.com:443/wd/hub",
                username, accessKey);

        try {
            WebDriver driver = new RemoteWebDriver(new URL(sauceUrl), options);
            driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(ConfigReader.getImplicitWait()));
            driver.manage().window().maximize();
            return driver;
        } catch (MalformedURLException e) {
            throw new RuntimeException("Invalid Sauce Labs URL: " + e.getMessage());
        }
    }

    /**
     * Update the Sauce Labs job result (pass/fail).
     */
    public static void updateSauceResult(WebDriver driver, boolean passed) {
        if (ConfigReader.isSauceEnabled()) {
            String script = "sauce:job-result=" + (passed ? "passed" : "failed");
            ((RemoteWebDriver) driver).executeScript(script);
        }
    }

    /**
     * Finds the highest-version cached driver binary under the given base directory.
     * e.g. ~/.cache/selenium/msedgedriver/win64/<version>/msedgedriver.exe
     * Returns the full path to the .exe, or null if not found.
     */
    private static String findCachedDriver(String baseDirPath, String driverFileName) {
        java.io.File baseDir = new java.io.File(baseDirPath);
        if (!baseDir.exists() || !baseDir.isDirectory()) return null;

        java.io.File[] versionDirs = baseDir.listFiles(java.io.File::isDirectory);
        if (versionDirs == null || versionDirs.length == 0) return null;

        // Sort descending to pick highest version first
        java.util.Arrays.sort(versionDirs, (a, b) -> b.getName().compareTo(a.getName()));

        for (java.io.File versionDir : versionDirs) {
            java.io.File driver = new java.io.File(versionDir, driverFileName);
            if (driver.exists() && driver.canExecute()) {
                return driver.getAbsolutePath();
            }
        }
        return null;
    }

    /** Unused for edge/chrome but kept for firefox compatibility. */
    @SuppressWarnings("unused")
    private static void setLocalDriverPath(String driverName, String property, String basePath) {
        java.io.File base = new java.io.File(basePath);
        if (!base.exists()) return;
        java.io.File[] versions = base.listFiles(java.io.File::isDirectory);
        if (versions == null || versions.length == 0) return;
        java.util.Arrays.sort(versions, (a, b) -> b.getName().compareTo(a.getName()));
        for (java.io.File v : versions) {
            java.io.File exe = new java.io.File(v, driverName + ".exe");
            if (exe.exists()) { System.setProperty(property, exe.getAbsolutePath()); return; }
        }
    }
}
